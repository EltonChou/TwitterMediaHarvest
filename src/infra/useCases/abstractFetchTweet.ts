import type { Factory } from '#domain/factories/base'
import type {
  FetchTweet,
  FetchTweetCommand,
  TweetResult,
} from '#domain/useCases/fetchTweet'
import { FetchTweetError, ParseTweetError } from '#domain/useCases/fetchTweet'
import { Tweet, type TweetProps } from '#domain/valueObjects/tweet'
import { TweetMedia } from '#domain/valueObjects/tweetMedia'
import { TweetUser, type TweetUserProps } from '#domain/valueObjects/tweetUser'
import { toErrorResult, toSuccessResult } from '#utils/result'
import * as A from 'fp-ts/Array'
import * as E from 'fp-ts/Either'
import * as O from 'fp-ts/Option'
import * as TE from 'fp-ts/TaskEither'
import { apply, flow, pipe } from 'fp-ts/function'
import Joi, { type ValidationResult } from 'joi'

type ParseOptions = {
  targetTweetId: string
}

type TimelineInstruction = {
  type: 'TimelineAddEntries'
  entries?: InstructionEntry[]
}

type InstructionEntry = {
  entryId: string
  content: any
}

export type MakeHeaderParams = {
  bearerToken: string
  csrfToken: string
}

const tweetPartialPropsSchema = Joi.object<
  Omit<TweetProps, 'user' | 'videos' | 'images'>
>({
  id: Joi.string().required(),
  hashtags: Joi.array().items(Joi.string()),
  createdAt: Joi.date().required(),
})

const tweetUserPropsSchema = Joi.object<TweetUserProps, true>({
  displayName: Joi.string(),
  screenName: Joi.string(),
  userId: Joi.string(),
})

export abstract class FetchTweetBase implements FetchTweet {
  abstract identity: string
  protected bearerToken =
    'AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA'

  abstract makeEndpoint(tweetId: string): string
  abstract makeHeaders(params: MakeHeaderParams): Headers

  protected parseBodyWithOptions(options: ParseOptions) {
    return (body: any): Result<Tweet> => {
      if (Object.hasOwn(body, 'errors')) return toErrorResult(new FetchTweetError(404))

      const getResultFromBody = (body: any) =>
        pipe(
          O.some<TimelineInstruction[]>(
            body?.data?.threaded_conversation_with_injections_v2?.instructions
          ),
          E.fromOption(() => 'Failed to get instructions'),
          E.chain(
            flow(
              A.filter(i => i.type === 'TimelineAddEntries'),
              A.head,
              E.fromOption(() => 'Failed to get instruction')
            )
          ),
          E.chain(instruction =>
            Array.isArray(instruction.entries)
              ? E.right(instruction.entries)
              : E.left('Failed to get entries')
          ),
          E.chain(
            flow(
              A.filter(e => e.entryId.includes(options.targetTweetId)),
              A.head,
              E.fromOption(() => 'Failed to get entry')
            )
          ),
          E.chain(entry =>
            pipe(
              entry?.content?.itemContent?.tweet_results?.result.tweet ??
                entry?.content?.itemContent?.tweet_results?.result,
              E.fromNullable('Failed to get result')
            )
          )
        )

      const getTweetResultFromResult = (result: any) =>
        pipe(result?.legacy ?? result, E.fromNullable('Failed to get tweet result'))

      const getTweetPropsFromTweetResult = (tweetResult: any) =>
        pipe(
          tweetPartialPropsSchema.validate({
            id: tweetResult.rest_id ?? tweetResult.id_str,
            hashtags: parseHashtags(tweetResult?.entities),
            createdAt: new Date(tweetResult.created_at),
          }),
          validationResultToEither
        )

      const getUserPropsFromResult = (result: any) =>
        pipe(
          result?.core?.user_results?.result,
          E.fromNullable('Failed to get user result'),
          E.chain(userResult =>
            pipe(
              tweetUserPropsSchema.validate({
                displayName: userResult?.legacy?.name,
                screenName: userResult?.legacy?.screen_name,
                userId: userResult?.rest_id,
              }),
              validationResultToEither
            )
          )
        )

      const getMediaCollectionFromTweetResult = (tweetResult: any) =>
        pipe(tweetResult?.extended_entities ?? [], parseMedias)

      const parseTweetFromBody = (body: any): Result<Tweet> =>
        pipe(
          E.Do,
          E.bind('result', () => getResultFromBody(body)),
          E.bind('tweetResult', payload => getTweetResultFromResult(payload.result)),
          E.bind('userProps', payload => getUserPropsFromResult(payload.result)),
          E.bind('mediaCollection', payload =>
            E.tryCatch(
              () => getMediaCollectionFromTweetResult(payload.tweetResult),
              e => E.toError(e).message ?? 'Failed to parse media collection'
            )
          ),
          E.bind('partialTweetProps', payload =>
            getTweetPropsFromTweetResult(payload.tweetResult)
          ),
          E.map(payload => ({
            user: new TweetUser(payload.userProps),
            ...payload.mediaCollection,
            ...payload.partialTweetProps,
          })),
          E.map(props => new Tweet(props)),
          E.mapLeft(r => new ParseTweetError(r)),
          E.match(toErrorResult, toSuccessResult)
        )

      return parseTweetFromBody(body)
    }
  }

  async process(command: FetchTweetCommand): Promise<TweetResult> {
    const callTweetApi = TE.tryCatch(
      () =>
        fetch(this.makeEndpoint(command.tweetId), {
          method: 'GET',
          headers: this.makeHeaders({
            bearerToken: this.bearerToken,
            csrfToken: command.csrfToken,
          }),
          mode: 'cors',
          referrer: `https://x.com/i/web/status/${command.tweetId}`,
        }),
      E.toError
    )

    const parseResponse = (resp: Response) =>
      pipe(
        TE.Do,
        TE.bind('quota', () =>
          pipe(
            resp.headers.get('X-Rate-Limit-Remaining'),
            O.fromNullable,
            O.orElse(() => O.some('0')),
            TE.fromOption(() => 'Failed to get rate limit remaining'),
            TE.mapLeft(E.toError),
            TE.map(Number.parseInt)
          )
        ),
        TE.bind('body', () => TE.tryCatch(() => resp.json(), E.toError)),
        TE.bind('tweet', ({ body }) =>
          pipe(
            this.parseBodyWithOptions,
            apply({ targetTweetId: command.tweetId }),
            apply(body),
            result => (result.error ? TE.left(result.error) : TE.right(result.value))
          )
        ),
        TE.map(
          ({ tweet, quota }) =>
            ({
              error: undefined,
              value: tweet,
              remainingQuota: quota,
            } satisfies TweetResult)
        )
      )

    const fetchTweet = pipe(
      callTweetApi,
      TE.chain(resp =>
        resp.status === 200 ? TE.right(resp) : TE.left(new FetchTweetError(resp.status))
      ),
      TE.flatMap(flow(parseResponse)),
      TE.match(flow(E.toError, errorToErrorTweetResult), r => r)
    )

    return fetchTweet()
  }
}

type MediaCollection = { images: TweetMedia[]; videos: TweetMedia[] }

const parseMedias = (entity: Record<string, unknown>): MediaCollection => {
  const medias = (entity?.media as Media[]) ?? []
  let imageIndex = 0
  let videoIndex = 0

  const increaseImageIdx = () => (imageIndex += 1)
  const increaseVideoIndex = () => (videoIndex += 1)

  return medias.reduce<MediaCollection>(
    (mediaCollection, media) => {
      mediaCollection.images.push(
        new TweetMedia({
          index: imageIndex,
          type: media.type === 'photo' ? 'photo' : 'thumbnail',
          url: media.media_url_https,
        })
      )
      increaseImageIdx()

      if (media.type === 'animated_gif' || media.type === 'video') {
        const url = parseBestVideoVariant(media.video_info.variants)
        if (url) {
          mediaCollection.videos.push(
            new TweetMedia({
              index: videoIndex,
              type: 'video',
              url: url,
            })
          )
          increaseVideoIndex()
        }
      }

      return mediaCollection
    },
    {
      images: [],
      videos: [],
    }
  )
}

type Hashtag = {
  text: string
}

const parseHashtags = (entity: Record<string, unknown>): string[] =>
  ((entity?.hashtags as Hashtag[]) ?? []).map(v => v.text)

type Mp4Variant = {
  bitrate: number
  content_type: 'video/mp4'
  url: string
}

type MpegUrlVariant = {
  bitrate: 0
  content_type: 'application/x-mpegURL'
  url: string
}

type VideoVariant = Mp4Variant | MpegUrlVariant

type Media =
  | {
      type: 'photo'
      media_url_https: string
    }
  | {
      type: 'video' | 'animated_gif'
      media_url_https: string
      video_info: {
        variants: VideoVariant[]
      }
    }

const parseBestVideoVariant = (variants: VideoVariant[]): string | undefined =>
  variants
    .filter(variant => variant.content_type === 'video/mp4')
    .reduce((prevVariant, currVariant) =>
      currVariant?.bitrate >= prevVariant?.bitrate ? currVariant : prevVariant
    ).url

const validationResultToEither = <T>(result: ValidationResult<T>): E.Either<string, T> =>
  result.error ? E.left(result.error.message) : E.right(result.value)

const errorToErrorTweetResult: Factory<Error, TweetResult> = error => ({
  error: error,
  remainingQuota: -1,
  value: undefined,
})
