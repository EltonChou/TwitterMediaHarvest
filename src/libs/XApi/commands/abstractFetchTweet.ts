import { Tweet, type TweetProps } from '#domain/valueObjects/tweet'
import { TweetMedia } from '#domain/valueObjects/tweetMedia'
import { TweetUser, type TweetUserProps } from '#domain/valueObjects/tweetUser'
import { toErrorResult, toSuccessResult } from '#utils/result'
import { GraphQLCommand, Query } from './graphql'
import { HttpMethod } from './types'
import type {
  CacheAble,
  Command,
  CommandCache,
  MetadataBearer,
  RequestContext,
  ResponseMetadata,
} from './types'
import * as A from 'fp-ts/Array'
import * as E from 'fp-ts/Either'
import * as O from 'fp-ts/Option'
import { flow, pipe } from 'fp-ts/function'
import Joi, { type ValidationResult } from 'joi'

export interface FetchTweetCommandInput extends LiteralObject {
  csrfToken: string
  tweetId: string
  cacheProvider: CommandCache | Provider<CommandCache>
}

// TODO: Should we isolate tweet object?
export interface FetchTweetCommandOutput extends MetadataBearer {
  tweetResult: Result<Tweet>
}

type TimelineInstruction = {
  type: 'TimelineAddEntries'
  entries?: InstructionEntry[]
}

type InstructionEntry = {
  entryId: string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  content: any
}

const tweetPartialPropsSchema = Joi.object<
  Omit<TweetProps, 'user' | 'videos' | 'images'>
>({
  id: Joi.string().required(),
  hashtags: Joi.array().items(Joi.string()).required(),
  createdAt: Joi.date().required().required(),
})

const tweetUserPropsSchema = Joi.object<TweetUserProps, true>({
  displayName: Joi.string().required(),
  screenName: Joi.string().required(),
  userId: Joi.string().required(),
  isProtected: Joi.boolean().default(false),
})

export abstract class FetchTweetCommand
  extends GraphQLCommand
  implements Command<FetchTweetCommandInput, FetchTweetCommandOutput>, CacheAble
{
  readonly isCacheAble = true
  readonly method = HttpMethod.Get
  readonly config: FetchTweetCommandInput
  private cache: CommandCache | undefined = undefined

  constructor(config: FetchTweetCommandInput, query: Query) {
    super(query)
    this.config = config
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  protected getResultFromBody(body: any) {
    return pipe(
      body?.data?.threaded_conversation_with_injections_v2?.instructions,
      O.fromNullable<TimelineInstruction[]>,
      E.fromOption(() => 'Failed to get instructions'),
      E.chain(
        flow(
          A.filter(i => i.type === 'TimelineAddEntries'),
          A.head,
          E.fromOption(() => 'Failed to get `TimelineAddEntries` sinstruction')
        )
      ),
      E.chain(instruction =>
        Array.isArray(instruction.entries)
          ? E.right(instruction.entries)
          : E.left('Failed to get entries')
      ),
      E.chain(
        flow(
          A.filter(e => e.entryId.includes(this.config.tweetId)),
          A.head,
          E.fromOption(() => 'Failed to get entry')
        )
      ),
      E.chain(entry =>
        pipe(
          entry?.content?.itemContent?.tweet_results?.result.tweet ??
            entry?.content?.itemContent?.tweet_results?.result,
          E.fromNullable('Failed to get result')
        )
      )
    )
  }

  protected parseBody(body: unknown): Result<Tweet> {
    if (!body) {
      return hasErrorProperty(body)
        ? toErrorResult(
            new ParseTweetError(body.error ?? body.errors ?? 'Invalid body')
          )
        : toErrorResult(new ParseTweetError('Invalid body'))
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const getTweetResultFromResult = (result: any) =>
      pipe(
        result?.legacy ?? result,
        E.fromNullable('Failed to get tweet result')
      )

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const getTweetPropsFromTweetResult = (tweetResult: any) =>
      pipe(
        tweetPartialPropsSchema.validate({
          id: tweetResult.rest_id ?? tweetResult.id_str,
          hashtags: parseHashtags(tweetResult?.entities),
          createdAt: new Date(tweetResult.created_at),
        }),
        validationResultToEither
      )

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const getUserPropsFromResult = (result: any) =>
      pipe(
        result?.core?.user_results?.result,
        E.fromNullable('Failed to get user result'),
        E.chain(userResult =>
          pipe(
            tweetUserPropsSchema.validate({
              isProtected: userResult?.legacy?.protected,
              displayName: userResult?.legacy?.name,
              screenName: userResult?.legacy?.screen_name,
              userId: userResult?.rest_id,
            }),
            validationResultToEither
          )
        )
      )

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const parseTweetFromBody = (body: any): Result<Tweet> =>
      pipe(
        E.Do,
        E.bind('result', () => this.getResultFromBody(body)),
        E.bind('tweetResult', payload =>
          getTweetResultFromResult(payload.result)
        ),
        E.bind('userProps', payload => getUserPropsFromResult(payload.result)),
        E.bind('medias', payload =>
          pipe(
            payload.tweetResult?.extended_entities?.media,
            E.fromNullable('Failed to get medias.')
          )
        ),
        E.bind('mediaCollection', payload =>
          E.tryCatch(
            () => parseMedias(payload.medias),
            e => E.toError(e).message ?? 'Failed to parse medias'
          )
        ),
        E.bind('partialTweetProps', payload =>
          getTweetPropsFromTweetResult(payload.tweetResult)
        ),
        E.map(payload => ({
          user: new TweetUser(payload.userProps),
          ...payload.mediaCollection,
          ...payload.partialTweetProps,
        })),
        E.map(props => new Tweet(props)),
        E.mapLeft(r => new ParseTweetError(r)),
        E.match(toErrorResult<Tweet>, toSuccessResult<Tweet>)
      )

    return parseTweetFromBody(body)
  }

  protected parseMetadata(response: Response): ResponseMetadata {
    const quota = pipe(
      response.headers.get('X-Rate-Limit-Remaining'),
      O.fromNullable,
      O.match<string, ResponseMetadata['remainingQuota']>(
        /**
         * If the header is not present, it might be because the request was not successful.
         * In this case, we return -1 to indicate that the quota is unknown.
         */
        () => undefined,
        q => Number.parseInt(q, 10)
      )
    )

    const resetTime = pipe(
      // The value is in seconds
      response.headers.get('X-Rate-Limit-Reset'),
      O.fromNullable,
      O.match<string, ResponseMetadata['quotaResetTime']>(
        () => undefined,
        ts => new Date(Number.parseInt(ts, 10) * 1000)
      )
    )

    return {
      httpStatusCode: response.status,
      remainingQuota: quota,
      quotaResetTime: resetTime,
    }
  }

  protected async getCache() {
    return (this.cache ??=
      typeof this.config.cacheProvider === 'function'
        ? await this.config.cacheProvider()
        : this.config.cacheProvider)
  }

  async readFromCache(request: Request): Promise<Response | undefined> {
    const cache = await this.getCache()
    return cache.get(request)
  }

  async putIntoCache(request: Request, response: Response): Promise<void> {
    const cache = await this.getCache()
    if (response.ok) return cache.put(request.clone(), response.clone())
  }

  prepareRequest(context: RequestContext): Request {
    return new Request(this.makeEndpoint(context), {
      method: this.method,
      headers: this.makeHeaders(this.makeAuthHeaders(this.config.csrfToken)),
      mode: 'cors',
      referrer: `https://x.com/i/web/status/${this.config.tweetId}`,
    })
  }

  async resolveResponse(response: Response): Promise<FetchTweetCommandOutput> {
    const metadata = this.parseMetadata(response)

    if (!response.ok)
      return {
        $metadata: metadata,
        tweetResult: toErrorResult(
          new FetchTweetError('Failed to fetch tweet', response.status)
        ),
      }

    try {
      const body = await response.json()
      const result = this.parseBody(body)
      return {
        $metadata: metadata,
        tweetResult: result,
      }
    } catch (error) {
      return {
        $metadata: metadata,
        tweetResult: toErrorResult(
          new ParseTweetError('Failed to parse body', { cause: error })
        ),
      }
    }
  }
}

type Hashtag = {
  text: string
}

const parseHashtags = (entity: Record<string, unknown>): string[] =>
  ((entity?.hashtags as Hashtag[]) ?? []).map(v => v.text)

const validationResultToEither = <T>(
  result: ValidationResult<T>
): E.Either<string, T> =>
  result.error ? E.left(result.error.message) : E.right(result.value)

type Mp4Variant = {
  bitrate: number
  content_type: 'video/mp4'
  url: string
}

type MpegUrlVariant = {
  bitrate: 0
  content_type: 'application/x-mpegURL'
  url: string
}

type VideoVariant = Mp4Variant | MpegUrlVariant

type Media =
  | {
      type: 'photo'
      media_url_https: string
    }
  | {
      type: 'video' | 'animated_gif'
      media_url_https: string
      video_info: {
        variants: VideoVariant[]
      }
    }

type MediaCollection = { images: TweetMedia[]; videos: TweetMedia[] }

const parseMedias = (medias: Media[]): MediaCollection => {
  let imageIndex = 0
  let videoIndex = 0

  const increaseImageIdx = () => (imageIndex += 1)
  const increaseVideoIndex = () => (videoIndex += 1)

  return medias.reduce<MediaCollection>(
    (mediaCollection, media) => {
      mediaCollection.images.push(
        new TweetMedia({
          index: imageIndex,
          type: media.type === 'photo' ? 'photo' : 'thumbnail',
          url: media.media_url_https,
        })
      )
      increaseImageIdx()

      if (media.type === 'animated_gif' || media.type === 'video') {
        const url = parseBestVideoVariant(media.video_info.variants)
        if (url) {
          mediaCollection.videos.push(
            new TweetMedia({
              index: videoIndex,
              type: 'video',
              url: url,
            })
          )
          increaseVideoIndex()
        }
      }

      return mediaCollection
    },
    {
      images: [],
      videos: [],
    }
  )
}

const parseBestVideoVariant = (variants: VideoVariant[]): string | undefined =>
  variants
    .filter(variant => variant.content_type === 'video/mp4')
    .reduce((prevVariant, currVariant) =>
      currVariant?.bitrate >= prevVariant?.bitrate ? currVariant : prevVariant
    ).url

const hasErrorProperty = (
  body: unknown
): body is { error?: string; errors?: string } => {
  return (
    typeof body === 'object' &&
    body !== null &&
    ('error' in body || 'errors' in body)
  )
}

export class ParseTweetError extends Error {
  name = 'ParseTweetError'
}

export class FetchTweetError extends Error {
  name = 'FetchTweetError'
  readonly statusCode: number

  constructor(message: string, statusCode: number, cause?: Error) {
    super(message, { cause })
    this.statusCode = statusCode
  }
}
