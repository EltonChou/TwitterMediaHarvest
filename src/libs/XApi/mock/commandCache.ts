import type { CommandCache } from '../commands/types'

export class MockCommandCache implements CommandCache {
  private readonly cache: Map<string, Response> = new Map()

  async get(request: Request): Promise<Response | undefined> {
    return this.cache.get(request.url)
  }

  async put(request: Request, response: Response): Promise<void> {
    this.cache.set(request.url, response)
  }
}
