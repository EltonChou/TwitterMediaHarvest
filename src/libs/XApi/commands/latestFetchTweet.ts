import { FetchTweetCommand } from './abstractFetchTweet'
import type { FetchTweetCommandInput } from './abstractFetchTweet'
import { AuthType, QueryMethod } from './graphql'

const featureParams = {
  rweb_tipjar_consumption_enabled: false,
  responsive_web_graphql_exclude_directive_enabled: false,
  verified_phone_label_enabled: false,
  creator_subscriptions_tweet_preview_api_enabled: false,
  responsive_web_graphql_timeline_navigation_enabled: false,
  responsive_web_graphql_skip_user_profile_image_extensions_enabled: false,
  communities_web_enable_tweet_community_results_fetch: false,
  c9s_tweet_anatomy_moderator_badge_enabled: false,
  articles_preview_enabled: true,
  tweetypie_unmention_optimization_enabled: false,
  responsive_web_edit_tweet_api_enabled: false,
  graphql_is_translatable_rweb_tweet_is_translatable_enabled: false,
  view_counts_everywhere_api_enabled: false,
  longform_notetweets_consumption_enabled: false,
  responsive_web_twitter_article_tweet_consumption_enabled: false,
  tweet_awards_web_tipping_enabled: false,
  creator_subscriptions_quote_tweet_preview_enabled: false,
  freedom_of_speech_not_reach_fetch_enabled: false,
  standardized_nudges_misinfo: false,
  tweet_with_visibility_results_prefer_gql_limited_actions_policy_enabled:
    false,
  tweet_with_visibility_results_prefer_gql_media_interstitial_enabled: false,
  rweb_video_timestamps_enabled: false,
  longform_notetweets_rich_text_read_enabled: false,
  longform_notetweets_inline_media_enabled: false,
  responsive_web_enhance_cards_enabled: false,
}

const fieldTogglesParams = {
  withArticleRichContentState: false,
  withAuxiliaryUserLabels: false,
}

export class LatestFetchTweetCommand extends FetchTweetCommand {
  readonly authType = AuthType.Auth
  constructor(config: FetchTweetCommandInput) {
    super(config, {
      id: 'zJvfJs3gSbrVhC0MKjt_OQ',
      name: 'TweetDetail',
      method: QueryMethod.Get,
      params: {
        variable: {
          focalTweetId: config.tweetId,
          with_rux_injections: false,
          includePromotedContent: false,
          withCommunity: false,
          withQuickPromoteEligibilityTweetFields: false,
          withBirdwatchNotes: false,
          withVoice: false,
          withV2Timeline: true,
        },
        feature: featureParams,
        fieldToggles: fieldTogglesParams,
      },
    })
  }
}
