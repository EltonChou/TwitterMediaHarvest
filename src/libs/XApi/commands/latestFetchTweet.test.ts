import { MockCommandCache } from '../mock/commandCache'
import { FetchTweetError, ParseTweetError } from './abstractFetchTweet'
import { LatestFetchTweetCommand } from './latestFetchTweet'
import fs from 'node:fs'
import path from 'node:path'

describe('unit test for latest fetch tweet command', () => {
  const mockResponseBody = JSON.parse(
    fs.readFileSync(
      path.resolve(
        __dirname,
        '..',
        '..',
        '..',
        'test-data',
        'TweetDetail.json'
      ),
      { encoding: 'utf-8' }
    )
  )

  const mockCache = new MockCommandCache()

  it.each([
    { cacheProvider: async () => mockCache },
    { cacheProvider: mockCache },
  ])('can cache', async ({ cacheProvider }) => {
    const command = new LatestFetchTweetCommand({
      tweetId: '1901057939403608167',
      csrfToken: 'token',
      cacheProvider: cacheProvider,
    })

    const request = command.prepareRequest({
      hostname: 'x.com',
      protocol: 'https',
    })

    const response = new Response(mockResponseBody, {
      headers: new Headers({ 'x-rate-limit-remaining': '100' }),
      status: 200,
    })

    await command.putIntoCache(request, response)

    const cachedResponse = await command.readFromCache(request)
    expect(cachedResponse).toBeDefined()
  })

  it('can prepare request', () => {
    const command = new LatestFetchTweetCommand({
      tweetId: '1901057939403608167',
      csrfToken: 'token',
      cacheProvider: mockCache,
    })

    const request = command.prepareRequest({
      hostname: 'x.com',
      protocol: 'https',
    })
    expect(request instanceof Request).toBeTrue()
  })

  it.each([
    { tweetId: '1901057939403608167', responseType: 'image' },
    { tweetId: '1900881067713982967', responseType: 'video' },
    { tweetId: '1829835601941823508', responseType: 'tagged-video' },
  ])(
    'should return a successful result when the $responseType response is ok',
    async ({ responseType, tweetId }) => {
      const command = new LatestFetchTweetCommand({
        tweetId: tweetId,
        csrfToken: 'token',
        cacheProvider: mockCache,
      })

      const resetTimeInSecond = 18888888

      const result = await command.resolveResponse(
        new Response(JSON.stringify(mockResponseBody[responseType]), {
          headers: new Headers({
            'x-rate-limit-remaining': '100',
            'x-rate-limit-reset': resetTimeInSecond.toString(),
          }),
          status: 200,
        })
      )

      expect(result).toMatchSnapshot()
      expect(result.$metadata.httpStatusCode).toBe(200)
      expect(result.$metadata.remainingQuota).toBe(100)
      expect(result.$metadata.quotaResetTime).toEqual(
        new Date(resetTimeInSecond * 1000)
      )
      expect(result.tweetResult.error).toBeUndefined()
      expect(result.tweetResult.value).toBeDefined()
    }
  )

  it('should handle a parse error when body is not json', async () => {
    const command = new LatestFetchTweetCommand({
      tweetId: '1901057939403608167',
      csrfToken: 'token',
      cacheProvider: mockCache,
    })

    const invalidBodyResponse = 'Invalid body'
    const resetTimeInSecond = 18888888

    const result = await command.resolveResponse(
      new Response(invalidBodyResponse, {
        headers: new Headers({
          'x-rate-limit-remaining': '100',
          'x-rate-limit-reset': resetTimeInSecond.toString(),
        }),
        status: 200,
      })
    )

    expect(result).toMatchSnapshot()
    expect(result.$metadata.httpStatusCode).toBe(200)
    expect(result.$metadata.remainingQuota).toBe(100)
    expect(result.$metadata.quotaResetTime).toEqual(
      new Date(resetTimeInSecond * 1000)
    )
    expect(result.tweetResult.error instanceof ParseTweetError).toBeTrue()
    expect(result.tweetResult.value).toBeUndefined()
  })

  it('should handle a parse error', async () => {
    const command = new LatestFetchTweetCommand({
      tweetId: '1901057939403608167',
      csrfToken: 'token',
      cacheProvider: mockCache,
    })

    const invalidJsonResponse = JSON.stringify({ error: 'invalid json' })
    const resetTimeInSecond = 18888888

    const result = await command.resolveResponse(
      new Response(invalidJsonResponse, {
        headers: new Headers({
          'x-rate-limit-remaining': '100',
          'x-rate-limit-reset': resetTimeInSecond.toString(),
        }),
        status: 200,
      })
    )

    expect(result).toMatchSnapshot()
    expect(result.$metadata.httpStatusCode).toBe(200)
    expect(result.$metadata.remainingQuota).toBe(100)
    expect(result.$metadata.quotaResetTime).toEqual(
      new Date(resetTimeInSecond * 1000)
    )
    expect(result.tweetResult.error instanceof ParseTweetError).toBeTrue()
    expect(result.tweetResult.value).toBeUndefined()
  })

  it('should handle FetchTweetError', async () => {
    const command = new LatestFetchTweetCommand({
      tweetId: '1901057939403608167',
      csrfToken: 'token',
      cacheProvider: mockCache,
    })

    const errorResponse = JSON.stringify({ error: 'Tweet not found' })
    const resetTimeInSecond = 18888888

    const result = await command.resolveResponse(
      new Response(errorResponse, {
        headers: new Headers({
          'x-rate-limit-remaining': '100',
          'x-rate-limit-reset': resetTimeInSecond.toString(),
        }),
        status: 404,
      })
    )

    expect(result).toMatchSnapshot()
    expect(result.$metadata.httpStatusCode).toBe(404)
    expect(result.$metadata.remainingQuota).toBe(100)
    expect(result.$metadata.quotaResetTime).toEqual(
      new Date(resetTimeInSecond * 1000)
    )
    expect(result.tweetResult.error instanceof FetchTweetError).toBeTrue()
    expect(result.tweetResult.value).toBeUndefined()
  })
})
