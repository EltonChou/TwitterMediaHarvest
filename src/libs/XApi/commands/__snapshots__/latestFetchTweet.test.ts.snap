// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test for latest fetch tweet command should handle FetchTweetError 1`] = `
{
  "$metadata": {
    "httpStatusCode": 404,
    "quotaResetTime": 1970-08-07T14:54:48.000Z,
    "remainingQuota": 100,
  },
  "tweetResult": {
    "error": [FetchTweetError: Failed to fetch tweet],
    "value": undefined,
  },
}
`;

exports[`unit test for latest fetch tweet command should handle a parse error 1`] = `
{
  "$metadata": {
    "httpStatusCode": 200,
    "quotaResetTime": 1970-08-07T14:54:48.000Z,
    "remainingQuota": 100,
  },
  "tweetResult": {
    "error": [ParseTweetError: invalid json],
    "value": undefined,
  },
}
`;

exports[`unit test for latest fetch tweet command should handle a parse error when body is not json 1`] = `
{
  "$metadata": {
    "httpStatusCode": 200,
    "quotaResetTime": 1970-08-07T14:54:48.000Z,
    "remainingQuota": 100,
  },
  "tweetResult": {
    "error": [ParseTweetError: Failed to parse body],
    "value": undefined,
  },
}
`;

exports[`unit test for latest fetch tweet command should return a successful result when the image response is ok 1`] = `
{
  "$metadata": {
    "httpStatusCode": 200,
    "quotaResetTime": 1970-08-07T14:54:48.000Z,
    "remainingQuota": 100,
  },
  "tweetResult": {
    "error": undefined,
    "value": {
      "createdAt": 2025-03-15T23:48:27.000Z,
      "hashtags": [],
      "id": "1901057939403608167",
      "images": [
        {
          "index": 0,
          "type": "photo",
          "url": "https://pbs.twimg.com/media/GmHqHpmWsAAuk_y.jpg",
        },
      ],
      "user": {
        "displayName": "Elon Musk",
        "isProtected": false,
        "screenName": "elonmusk",
        "userId": "44196397",
      },
      "videos": [],
    },
  },
}
`;

exports[`unit test for latest fetch tweet command should return a successful result when the tagged-video response is ok 1`] = `
{
  "$metadata": {
    "httpStatusCode": 200,
    "quotaResetTime": 1970-08-07T14:54:48.000Z,
    "remainingQuota": 100,
  },
  "tweetResult": {
    "error": undefined,
    "value": {
      "createdAt": 2024-08-31T10:56:18.000Z,
      "hashtags": [
        "‰ªäÊúàÊèè„ÅÑ„ÅüÁµµ„ÇíÊôí„Åù„ÅÜ",
      ],
      "id": "1829835601941823508",
      "images": [
        {
          "index": 0,
          "type": "thumbnail",
          "url": "https://pbs.twimg.com/tweet_video_thumb/GWThuAObQAEwO3_.jpg",
        },
        {
          "index": 1,
          "type": "thumbnail",
          "url": "https://pbs.twimg.com/tweet_video_thumb/GWThvdZbQAI2bgw.jpg",
        },
      ],
      "user": {
        "displayName": "birdmanü¶ä",
        "isProtected": false,
        "screenName": "birdman46049238",
        "userId": "1287977244023914496",
      },
      "videos": [
        {
          "index": 0,
          "type": "video",
          "url": "https://video.twimg.com/tweet_video/GWThuAObQAEwO3_.mp4",
        },
        {
          "index": 1,
          "type": "video",
          "url": "https://video.twimg.com/tweet_video/GWThvdZbQAI2bgw.mp4",
        },
      ],
    },
  },
}
`;

exports[`unit test for latest fetch tweet command should return a successful result when the video response is ok 1`] = `
{
  "$metadata": {
    "httpStatusCode": 200,
    "quotaResetTime": 1970-08-07T14:54:48.000Z,
    "remainingQuota": 100,
  },
  "tweetResult": {
    "error": undefined,
    "value": {
      "createdAt": 2025-03-15T12:05:37.000Z,
      "hashtags": [],
      "id": "1900881067713982967",
      "images": [
        {
          "index": 0,
          "type": "thumbnail",
          "url": "https://pbs.twimg.com/tweet_video_thumb/GmFIv_7bcAM-RZa.jpg",
        },
      ],
      "user": {
        "displayName": "birdmanü¶ä",
        "isProtected": false,
        "screenName": "birdman46049238",
        "userId": "1287977244023914496",
      },
      "videos": [
        {
          "index": 0,
          "type": "video",
          "url": "https://video.twimg.com/tweet_video/GmFIv_7bcAM-RZa.mp4",
        },
      ],
    },
  },
}
`;
